crates/amaru-ledger/src/rules.rs

```diff
@@ -43,29 +63,87 @@ pub fn validate_block(
         .map_err(|err| BlockValidationError::RuleViolations(vec![err.into()]))?;
 
     // TODO: rewrite this to use iterators defined on `Redeemers` and `MaybeIndefArray`, ideally
-    let ex_units = match block.transaction_witness_sets {
-        MaybeIndefArray::Def(vec) => vec,
-        MaybeIndefArray::Indef(vec) => vec,
-    }
-    .into_iter()
-    .flat_map(|witness_set| {
-        witness_set
-            .redeemer
-            .into_iter()
-            .map(|redeemers| match redeemers {
-                Redeemers::List(list) => list.iter().map(|r| r.ex_units).collect::<Vec<_>>(),
-                Redeemers::Map(map) => map.iter().map(|(_, r)| r.ex_units).collect::<Vec<_>>(),
-            })
-    })
-    .flatten()
-    .collect::<Vec<_>>();
+    let ex_units = block
+        .transaction_witness_sets
+        .clone()
+        .to_vec()
+        .into_iter()
+        .flat_map(|witness_set| {
+            witness_set
+                .redeemer
+                .into_iter()
+                .map(|redeemers| match redeemers {
+                    Redeemers::List(list) => list.iter().map(|r| r.ex_units).collect::<Vec<_>>(),
+                    Redeemers::Map(map) => map.iter().map(|(_, r)| r.ex_units).collect::<Vec<_>>(),
+                })
+        })
+        .flatten()
+        .collect::<Vec<_>>();
 
     block_ex_units_valid(ex_units, &protocol_params)
         .map_err(|err| BlockValidationError::RuleViolations(vec![err.into()]))?;
 
+    let transactions = block.transaction_bodies.to_vec();
+
+    let raw_transactions = minted_block.transaction_bodies.clone().to_vec();
+
+    let witness_sets = block.transaction_witness_sets.to_vec();
+
+    let invalid_transactions = match block.invalid_transactions {
+        Some(invalid_transactions) => invalid_transactions.to_vec(),
+        None => Vec::new(),
+    };
+
+    for (i, (transaction, raw_transaction)) in
+        (0u32..).zip(transactions.iter().zip(raw_transactions))
+    {
```

Comment:
First `zip` might be replaced with `transactions.iter().enumerate` ?

Reply:
enumerate returns a usize, not a u32. Since I need to check the presence of a u32 in an array with `i`, I thought it was safer to use `zip` and cast from u32 to usize where needed.

Reply:
I think it's reasonable to expect less than 4 billions transactions in a block 😄 ... A comment to justify that sounds acceptable.